syntax = "proto3";

package domunity;

option go_package = "github.com/domunity/backend/proto";

// ==================== Authentication Service ====================
service AuthService {
  rpc Login(LoginRequest) returns (LoginResponse) {}
  rpc Register(RegisterRequest) returns (RegisterResponse) {}
  rpc RefreshToken(RefreshTokenRequest) returns (RefreshTokenResponse) {}
  rpc ForgotPassword(ForgotPasswordRequest) returns (ForgotPasswordResponse) {}
}

message LoginRequest {
  string email = 1;
  string password = 2;
}

message LoginResponse {
  bool success = 1;
  string message = 2;
  string access_token = 3;
  string refresh_token = 4;
  User user = 5;
}

message RegisterRequest {
  string email = 1;
  string password = 2;
  string full_name = 3;
  string phone = 4;
}

message RegisterResponse {
  bool success = 1;
  string message = 2;
  string user_id = 3;
}

message RefreshTokenRequest {
  string refresh_token = 1;
}

message RefreshTokenResponse {
  bool success = 1;
  string access_token = 2;
}

message ForgotPasswordRequest {
  string email = 1;
}

message ForgotPasswordResponse {
  bool success = 1;
  string message = 2;
}

// ==================== User Service ====================
service UserService {
  rpc GetProfile(GetProfileRequest) returns (UserProfile) {}
  rpc UpdateProfile(UpdateProfileRequest) returns (UpdateProfileResponse) {}
}

message GetProfileRequest {
  string user_id = 1;
}

message User {
  string id = 1;
  string email = 2;
  string full_name = 3;
  string phone = 4;
  string created_at = 5;
}

message UserProfile {
  User user = 1;
  Building building = 2;
  Apartment apartment = 3;
  string account_manager = 4;
  double balance = 5;
  string client_number = 6;
  string contract_end_date = 7;
}

message UpdateProfileRequest {
  string user_id = 1;
  string full_name = 2;
  string phone = 3;
}

message UpdateProfileResponse {
  bool success = 1;
  string message = 2;
}

// ==================== Building Service ====================
service BuildingService {
  rpc GetBuilding(GetBuildingRequest) returns (Building) {}
  rpc ListApartments(ListApartmentsRequest) returns (ListApartmentsResponse) {}
}

message GetBuildingRequest {
  string building_id = 1;
}

message Building {
  string id = 1;
  string address = 2;
  string entrance = 3;
  int32 total_apartments = 4;
  int32 total_residents = 5;
}

message Apartment {
  string id = 1;
  string building_id = 2;
  int32 number = 3;
  int32 floor = 4;
  string type = 5;
  int32 residents = 6;
}

message ListApartmentsRequest {
  string building_id = 1;
}

message ListApartmentsResponse {
  repeated Apartment apartments = 1;
}

// ==================== Financial Service ====================
service FinancialService {
  rpc GetFinancialReport(GetFinancialReportRequest) returns (FinancialReport) {}
  rpc GetPaymentHistory(GetPaymentHistoryRequest) returns (PaymentHistory) {}
}

message GetFinancialReportRequest {
  string user_id = 1;
  string building_id = 2;
}

message FinancialReportEntry {
  int32 apartment_number = 1;
  string type = 2;
  int32 floor = 3;
  string client_name = 4;
  int32 residents = 5;
  double elevator_gtp = 6;
  double elevator_electricity = 7;
  double common_area_electricity = 8;
  double elevator_maintenance = 9;
  double management_fee = 10;
  double repair_fund = 11;
  double total_due = 12;
}

message FinancialReport {
  repeated FinancialReportEntry entries = 1;
  double total_balance = 2;
}

message GetPaymentHistoryRequest {
  string user_id = 1;
}

message Payment {
  string id = 1;
  string date = 2;
  double amount = 3;
  string description = 4;
  string status = 5;
}

message PaymentHistory {
  repeated Payment payments = 1;
}

// ==================== Event Service ====================
service EventService {
  rpc ListEvents(ListEventsRequest) returns (ListEventsResponse) {}
  rpc CreateEvent(CreateEventRequest) returns (CreateEventResponse) {}
}

message ListEventsRequest {
  string building_id = 1;
  int32 limit = 2;
}

message Event {
  string id = 1;
  string date = 2;
  string title = 3;
  string description = 4;
  string building_id = 5;
}

message ListEventsResponse {
  repeated Event events = 1;
}

message CreateEventRequest {
  string building_id = 1;
  string date = 2;
  string title = 3;
  string description = 4;
}

message CreateEventResponse {
  bool success = 1;
  string message = 2;
  string event_id = 3;
}

// ==================== Contact Service ====================
service ContactService {
  rpc SendContactForm(ContactFormRequest) returns (ContactFormResponse) {}
  rpc RequestOffer(OfferRequest) returns (OfferResponse) {}
  rpc RequestPresentation(PresentationRequest) returns (PresentationResponse) {}
}

message ContactFormRequest {
  string name = 1;
  string phone = 2;
  string email = 3;
  string message = 4;
}

message ContactFormResponse {
  bool success = 1;
  string message = 2;
}

message OfferRequest {
  string phone = 1;
  string email = 2;
  string city = 3;
  int32 num_properties = 4;
  string address = 5;
  string additional_info = 6;
}

message OfferResponse {
  bool success = 1;
  string message = 2;
}

message PresentationRequest {
  string date = 1;
  string building_type = 2;
  string phone = 3;
  string email = 4;
  string address = 5;
  string additional_info = 6;
}

message PresentationResponse {
  bool success = 1;
  string message = 2;
}

// ==================== Health Check Service ====================
service HealthService {
  rpc Check(HealthCheckRequest) returns (HealthCheckResponse) {}
}

message HealthCheckRequest {}

message HealthCheckResponse {
  bool healthy = 1;
  string version = 2;
  string database_status = 3;
}
